File: backend/src/main/java/com/shikhi/practice/service/PracticeGenerator.java
package com.shikhi.practice.service;

import com.shikhi.content.domain.Vocabulary;
import com.shikhi.practice.domain.PracticeExercise;
import com.shikhi.practice.domain.PracticeExerciseType;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.Set;
import java.util.UUID;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.springframework.stereotype.Component;

/**
 * Turns picked vocabulary into concrete exercises (E12, LLD §2.8). Pure templating — no
 * I/O — so the mix rules are unit-testable: a round cycles word-level and sentence-level
 * formats (~60/40), and sentence formats are only chosen when the word's example sentence
 * is short (US-12.5, "smaller sentences"). Correctness goes into the server-only answer
 * key; the payload holds only what the learner may see.
 */
@Component
public class PracticeGenerator {

	/** Rounds cycle this sequence (word, sentence, word, sentence, word → 60/40 over 10). */
	private static final PracticeExerciseType[] TYPE_CYCLE = {
			PracticeExerciseType.WORD_MEANING, PracticeExerciseType.SENTENCE_GAP,
			PracticeExerciseType.MEANING_WORD, PracticeExerciseType.SENTENCE_BUILD,
			PracticeExerciseType.TYPE_WORD };

	/** SENTENCE_BUILD only for genuinely small sentences (word-bank stays tappable). */
	static final int MAX_BUILD_TOKENS = 8;
	private static final int MIN_BUILD_TOKENS = 3;
	private static final int MCQ_DISTRACTORS = 3;

	private final Random random;

	public PracticeGenerator() {
		this(new Random());
	}

	PracticeGenerator(Random random) {
		this.random = random;
	}

	/**
	 * Build one round of exercises for {@code words}, drawing MCQ distractors from
	 * {@code distractorsByBand} (same-band words, any size — filtered per exercise).
	 */
	public List<PracticeExercise> generateRound(UUID sessionId, int round, List<Vocabulary> words,
			Map<String, List<Vocabulary>> distractorsByBand) {
		List<PracticeExercise> exercises = new ArrayList<>(words.size());
		for (int i = 0; i < words.size(); i++) {
			Vocabulary word = words.get(i);
			PracticeExerciseType type = chooseType(TYPE_CYCLE[i % TYPE_CYCLE.length], word);
			List<Vocabulary> pool = distractorsByBand.getOrDefault(word.getCefrLevel(), List.of());
			exercises.add(build(sessionId, round, i + 1, word, type, pool));
		}
		return exercises;
	}

	/** Fall back to an always-eligible word-level format when a word can't carry the type. */
	private PracticeExerciseType chooseType(PracticeExerciseType desired, Vocabulary word) {
		return switch (desired) {
			case SENTENCE_GAP -> canGap(word) ? desired
					: canBuild(word) ? PracticeExerciseType.SENTENCE_BUILD
					: PracticeExerciseType.WORD_MEANING;
			case SENTENCE_BUILD -> canBuild(word) ? desired
					: canGap(word) ? PracticeExerciseType.SENTENCE_GAP
					: PracticeExerciseType.WORD_MEANING;
			case TYPE_WORD -> isTypeable(word) ? desired : PracticeExerciseType.MEANING_WORD;
			case WORD_MEANING, MEANING_WORD -> desired;
		};
	}

	private PracticeExercise build(UUID sessionId, int round, int ordinal, Vocabulary word,
			PracticeExerciseType type, List<Vocabulary> pool) {
		return switch (type) {
			case WORD_MEANING -> wordMeaning(sessionId, round, ordinal, word, pool);
			case MEANING_WORD -> meaningWord(sessionId, round, ordinal, word, pool);
			case SENTENCE_GAP -> sentenceGap(sessionId, round, ordinal, word, pool);
			case SENTENCE_BUILD -> sentenceBuild(sessionId, round, ordinal, word);
			case TYPE_WORD -> typeWord(sessionId, round, ordinal, word);
		};
	}

	// ---- word-level formats -------------------------------------------------------------

	private PracticeExercise wordMeaning(UUID sessionId, int round, int ordinal, Vocabulary word,
			List<Vocabulary> pool) {
		Options options = mcqOptions(word, pool, Vocabulary::getBnGloss);
		Map<String, Object> payload = new LinkedHashMap<>();
		payload.put("options", options.entries());
		return new PracticeExercise(sessionId, round, ordinal, word.getId(),
				PracticeExerciseType.WORD_MEANING,
				"What does “%s” mean?".formatted(word.getHeadword()),
				"“%s” শব্দের অর্থ কী?".formatted(word.getHeadword()),
				payload, mcqKey(options, word.getHeadword() + " — " + word.getBnGloss()));
	}

	private PracticeExercise meaningWord(UUID sessionId, int round, int ordinal, Vocabulary word,
			List<Vocabulary> pool) {
		Options options = mcqOptions(word, pool, Vocabulary::getHeadword);
		Map<String, Object> payload = new LinkedHashMap<>();
		payload.put("options", options.entries());
		return new PracticeExercise(sessionId, round, ordinal, word.getId(),
				PracticeExerciseType.MEANING_WORD,
				"Which English word means “%s”?".formatted(word.getBnGloss()),
				"“%s” — এর ইংরেজি শব্দ কোনটি?".formatted(word.getBnGloss()),
				payload, mcqKey(options, word.getHeadword() + " — " + word.getBnGloss()));
	}

	private PracticeExercise typeWord(UUID sessionId, int round, int ordinal, Vocabulary word) {
		Map<String, Object> payload = new LinkedHashMap<>();
		payload.put("partOfSpeech", word.getPartOfSpeech());
		Map<String, Object> key = new LinkedHashMap<>();
		key.put("accepted", List.of(word.getHeadword()));
		key.put("revealText", word.getHeadword() + " — " + word.getBnGloss());
		return new PracticeExercise(sessionId, round, ordinal, word.getId(),
				PracticeExerciseType.TYPE_WORD,
				"Type the English word for “%s”".formatted(word.getBnGloss()),
				"“%s” — এর ইংরেজি শব্দটি লিখুন".formatted(word.getBnGloss()),
				payload, key);
	}

	// ---- sentence-level formats (short sentences only) -----------------------------------

	private PracticeExercise sentenceGap(UUID sessionId, int round, int ordinal, Vocabulary word,
			List<Vocabulary> pool) {
		String blanked = headwordPattern(word).matcher(word.getExampleEn()).replaceFirst("____");
		Options options = mcqOptions(word, pool, Vocabulary::getHeadword);
		Map<String, Object> payload = new LinkedHashMap<>();
		payload.put("options", options.entries());
		payload.put("contextBn", word.getExampleBn());
		return new PracticeExercise(sessionId, round, ordinal, word.getId(),
				PracticeExerciseType.SENTENCE_GAP,
				"Fill in the blank: “%s”".formatted(blanked),
				"শূন্যস্থান পূরণ করুন: “%s”".formatted(blanked),
				payload, mcqKey(options, word.getExampleEn()));
	}

	private PracticeExercise sentenceBuild(UUID sessionId, int round, int ordinal,
			Vocabulary word) {
		String sentence = stripTerminalPunctuation(word.getExampleEn());
		List<String> tokens = new ArrayList<>(List.of(sentence.split("\\s+")));
		List<Map<String, Object>> tiles = new ArrayList<>(tokens.size());
		for (String token : tokens) {
			tiles.add(Map.of("id", UUID.randomUUID().toString(),
					"text", Map.of("en", token, "bn", token)));
		}
		java.util.Collections.shuffle(tiles, random);
		Map<String, Object> payload = new LinkedHashMap<>();
		payload.put("tokens", tiles);
		payload.put("targetBn", word.getExampleBn());
		Map<String, Object> key = new LinkedHashMap<>();
		key.put("accepted", List.of(sentence));
		key.put("revealText", word.getExampleEn());
		return new PracticeExercise(sessionId, round, ordinal, word.getId(),
				PracticeExerciseType.SENTENCE_BUILD,
				"Build the sentence: “%s”".formatted(word.getExampleBn()),
				"শব্দ সাজিয়ে বাক্যটি তৈরি করুন: “%s”".formatted(word.getExampleBn()),
				payload, key);
	}

	// ---- eligibility ----------------------------------------------------------------------

	private boolean canGap(Vocabulary word) {
		return word.getExampleEn() != null && word.getExampleBn() != null
				&& isTypeable(word)
				&& headwordPattern(word).matcher(word.getExampleEn()).find();
	}

	private boolean canBuild(Vocabulary word) {
		if (word.getExampleEn() == null || word.getExampleBn() == null) {
			return false;
		}
		int tokens = stripTerminalPunctuation(word.getExampleEn()).split("\\s+").length;
		return tokens >= MIN_BUILD_TOKENS && tokens <= MAX_BUILD_TOKENS;
	}

	/** Single-token headwords only ("advice", not "a, an") — anything a learner can type. */
	private boolean isTypeable(Vocabulary word) {
		return !word.getHeadword().contains(" ") && !word.getHeadword().contains(",");
	}

	private Pattern headwordPattern(Vocabulary word) {
		return Pattern.compile("\\b" + Pattern.quote(word.getHeadword()) + "\\b",
				Pattern.CASE_INSENSITIVE);
	}

	private String stripTerminalPunctuation(String sentence) {
		return sentence.replaceAll("[.!?।]+$", "").trim();
	}

	// ---- MCQ assembly ---------------------------------------------------------------------

	private record Options(List<Map<String, Object>> entries, String correctId) {
	}

	/**
	 * Correct value + up to three same-band distractors (same part of speech preferred,
	 * duplicate texts skipped), shuffled. The correct flag never enters the payload — only
	 * the answer key knows {@code correctId}.
	 */
	private Options mcqOptions(Vocabulary word, List<Vocabulary> pool,
			java.util.function.Function<Vocabulary, String> text) {
		List<Vocabulary> distractors = new ArrayList<>(MCQ_DISTRACTORS);
		Set<String> taken = new HashSet<>();
		taken.add(text.apply(word));
		// Two passes: same part of speech first, then anything still needed. A text only
		// enters `taken` when actually used, so pass 2 can still pick pass-1 rejects.
		for (boolean samePosOnly : new boolean[] { true, false }) {
			for (Vocabulary candidate : pool) {
				if (distractors.size() == MCQ_DISTRACTORS) {
					break;
				}
				if (candidate.getId().equals(word.getId())
						|| (samePosOnly && !candidate.getPartOfSpeech().equals(word.getPartOfSpeech()))
						|| !taken.add(text.apply(candidate))) {
					continue;
				}
				distractors.add(candidate);
			}
		}

		String correctId = UUID.randomUUID().toString();
		List<Map<String, Object>> entries = new ArrayList<>(distractors.size() + 1);
		entries.add(option(correctId, text.apply(word)));
		for (Vocabulary distractor : distractors) {
			entries.add(option(UUID.randomUUID().toString(), text.apply(distractor)));
		}
		java.util.Collections.shuffle(entries, random);
		return new Options(entries, correctId);
	}

	private Map<String, Object> option(String id, String text) {
		Map<String, Object> entry = new HashMap<>();
		entry.put("id", id);
		entry.put("text", Map.of("en", text, "bn", text));
		return entry;
	}

	private Map<String, Object> mcqKey(Options options, String revealText) {
		Map<String, Object> key = new LinkedHashMap<>();
		key.put("correctOptionId", options.correctId());
		key.put("revealText", revealText);
		return key;
	}
}

File: android/app/src/main/java/com/shikhi/app/data/practice/PracticeGenerator.kt
package com.shikhi.app.data.practice

import com.shikhi.app.data.content.db.LocalVocabulary
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.add
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray
import java.util.UUID
import kotlin.random.Random

/**
 * One generated exercise, persisted as a [com.shikhi.app.data.db.LocalPracticeExercise] row.
 * `payload` is the learner-visible config; `answerKey` is grading-only and must never reach the
 * UI layer directly (same "correctness never leaves the answer key" rule as the backend and as
 * OF3's [com.shikhi.app.data.content.db.ContentAnswerKeyDao] split) — see
 * [LocalPracticeSource.toWireExercise], which strips it when building the learner-facing DTO.
 */
data class GeneratedExercise(
	val round: Int,
	val ordinal: Int,
	val vocabularyId: String,
	val type: String,
	val promptEn: String,
	val promptBn: String,
	val payload: JsonObject,
	val answerKey: JsonObject,
)

/**
 * Verbatim-behavior Kotlin port of the backend's `PracticeGenerator`
 * (`backend/src/main/java/com/shikhi/practice/service/PracticeGenerator.java`, docs/
 * 93-offline-learning-design.md §4.3): pure templating, no I/O, so a round of exercises is fully
 * deterministic given its inputs and [random]. Prompt/feedback string templates (English and
 * Bengali) are copied verbatim — these are user-visible text.
 */
class PracticeGenerator(private val random: Random = Random.Default) {

	/** Rounds cycle this sequence (word, sentence, word, sentence, word -> 60/40 over 10). */
	private val typeCycle = listOf(
		PracticeExerciseType.WORD_MEANING,
		PracticeExerciseType.SENTENCE_GAP,
		PracticeExerciseType.MEANING_WORD,
		PracticeExerciseType.SENTENCE_BUILD,
		PracticeExerciseType.TYPE_WORD,
	)

	/**
	 * Build one round of exercises for [words], drawing MCQ distractors from
	 * [distractorsByBand] (same-band words, any size — filtered per exercise).
	 */
	fun generateRound(
		round: Int,
		words: List<LocalVocabulary>,
		distractorsByBand: Map<String, List<LocalVocabulary>>,
	): List<GeneratedExercise> =
		words.mapIndexed { i, word ->
			val type = chooseType(typeCycle[i % typeCycle.size], word)
			val pool = distractorsByBand[word.cefrLevel].orEmpty()
			build(round, i + 1, word, type, pool)
		}

	/** Fall back to an always-eligible word-level format when a word can't carry the type. */
	private fun chooseType(desired: String, word: LocalVocabulary): String = when (desired) {
		PracticeExerciseType.SENTENCE_GAP -> when {
			canGap(word) -> desired
			canBuild(word) -> PracticeExerciseType.SENTENCE_BUILD
			else -> PracticeExerciseType.WORD_MEANING
		}
		PracticeExerciseType.SENTENCE_BUILD -> when {
			canBuild(word) -> desired
			canGap(word) -> PracticeExerciseType.SENTENCE_GAP
			else -> PracticeExerciseType.WORD_MEANING
		}
		PracticeExerciseType.TYPE_WORD -> if (isTypeable(word)) desired else PracticeExerciseType.MEANING_WORD
		else -> desired // WORD_MEANING, MEANING_WORD
	}

	private fun build(
		round: Int,
		ordinal: Int,
		word: LocalVocabulary,
		type: String,
		pool: List<LocalVocabulary>,
	): GeneratedExercise = when (type) {
		PracticeExerciseType.WORD_MEANING -> wordMeaning(round, ordinal, word, pool)
		PracticeExerciseType.MEANING_WORD -> meaningWord(round, ordinal, word, pool)
		PracticeExerciseType.SENTENCE_GAP -> sentenceGap(round, ordinal, word, pool)
		PracticeExerciseType.SENTENCE_BUILD -> sentenceBuild(round, ordinal, word)
		PracticeExerciseType.TYPE_WORD -> typeWord(round, ordinal, word)
		else -> error("Unknown practice exercise type: $type")
	}

	// ---- word-level formats -------------------------------------------------------------

	private fun wordMeaning(round: Int, ordinal: Int, word: LocalVocabulary, pool: List<LocalVocabulary>): GeneratedExercise {
		val options = mcqOptions(word, pool) { it.bnGloss }
		val payload = buildJsonObject { put("options", options.toJson()) }
		return GeneratedExercise(
			round, ordinal, word.id, PracticeExerciseType.WORD_MEANING,
			promptEn = "What does “${word.headword}” mean?",
			promptBn = "“${word.headword}” শব্দের অর্থ কী?",
			payload = payload,
			answerKey = mcqKey(options, "${word.headword} — ${word.bnGloss}"),
		)
	}

	private fun meaningWord(round: Int, ordinal: Int, word: LocalVocabulary, pool: List<LocalVocabulary>): GeneratedExercise {
		val options = mcqOptions(word, pool) { it.headword }
		val payload = buildJsonObject { put("options", options.toJson()) }
		return GeneratedExercise(
			round, ordinal, word.id, PracticeExerciseType.MEANING_WORD,
			promptEn = "Which English word means “${word.bnGloss}”?",
			promptBn = "“${word.bnGloss}” — এর ইংরেজি শব্দ কোনটি?",
			payload = payload,
			answerKey = mcqKey(options, "${word.headword} — ${word.bnGloss}"),
		)
	}

	private fun typeWord(round: Int, ordinal: Int, word: LocalVocabulary): GeneratedExercise {
		val payload = buildJsonObject { put("partOfSpeech", word.partOfSpeech) }
		val key = buildJsonObject {
			putJsonArray("accepted") { add(word.headword) }
			put("revealText", "${word.headword} — ${word.bnGloss}")
		}
		return GeneratedExercise(
			round, ordinal, word.id, PracticeExerciseType.TYPE_WORD,
			promptEn = "Type the English word for “${word.bnGloss}”",
			promptBn = "“${word.bnGloss}” — এর ইংরেজি শব্দটি লিখুন",
			payload = payload,
			answerKey = key,
		)
	}

	// ---- sentence-level formats (short sentences only) -----------------------------------

	private fun sentenceGap(round: Int, ordinal: Int, word: LocalVocabulary, pool: List<LocalVocabulary>): GeneratedExercise {
		val exampleEn = requireNotNull(word.exampleEn)
		val blanked = headwordPattern(word).replaceFirst(exampleEn, "____")
		val options = mcqOptions(word, pool) { it.headword }
		val payload = buildJsonObject {
			put("options", options.toJson())
			put("contextBn", word.exampleBn)
		}
		return GeneratedExercise(
			round, ordinal, word.id, PracticeExerciseType.SENTENCE_GAP,
			promptEn = "Fill in the blank: “$blanked”",
			promptBn = "শূন্যস্থান পূরণ করুন: “$blanked”",
			payload = payload,
			answerKey = mcqKey(options, exampleEn),
		)
	}

	private fun sentenceBuild(round: Int, ordinal: Int, word: LocalVocabulary): GeneratedExercise {
		val sentence = stripTerminalPunctuation(requireNotNull(word.exampleEn))
		val tokens = sentence.split(Regex("\\s+"))
		val tiles = tokens.map { token -> TileId(UUID.randomUUID().toString(), token) }.shuffled(random)
		val payload = buildJsonObject {
			putJsonArray("tokens") {
				tiles.forEach { tile ->
					add(
						buildJsonObject {
							put("id", tile.id)
							put("textEn", tile.text)
							put("textBn", tile.text)
						},
					)
				}
			}
			put("targetBn", word.exampleBn)
		}
		val key = buildJsonObject {
			putJsonArray("accepted") { add(sentence) }
			put("revealText", word.exampleEn)
		}
		return GeneratedExercise(
			round, ordinal, word.id, PracticeExerciseType.SENTENCE_BUILD,
			promptEn = "Build the sentence: “${word.exampleBn}”",
			promptBn = "শব্দ সাজিয়ে বাক্যটি তৈরি করুন: “${word.exampleBn}”",
			payload = payload,
			answerKey = key,
		)
	}

	// ---- eligibility ----------------------------------------------------------------------

	private fun canGap(word: LocalVocabulary): Boolean {
		val exampleEn = word.exampleEn ?: return false
		if (word.exampleBn == null || !isTypeable(word)) return false
		return headwordPattern(word).containsMatchIn(exampleEn)
	}

	private fun canBuild(word: LocalVocabulary): Boolean {
		val exampleEn = word.exampleEn ?: return false
		if (word.exampleBn == null) return false
		val tokenCount = stripTerminalPunctuation(exampleEn).split(Regex("\\s+")).size
		return tokenCount in MIN_BUILD_TOKENS..MAX_BUILD_TOKENS
	}

	/** Single-token headwords only ("advice", not "a, an") — anything a learner can type. */
	private fun isTypeable(word: LocalVocabulary): Boolean =
		!word.headword.contains(" ") && !word.headword.contains(",")

	private fun headwordPattern(word: LocalVocabulary): Regex =
		Regex("\\b${Regex.escape(word.headword)}\\b", RegexOption.IGNORE_CASE)

	private fun stripTerminalPunctuation(sentence: String): String =
		sentence.replace(Regex("[.!?।]+$"), "").trim()

	// ---- MCQ assembly ---------------------------------------------------------------------

	private data class TileId(val id: String, val text: String)

	private data class OptionEntry(val id: String, val text: String)

	private data class Options(val entries: List<OptionEntry>, val correctId: String) {
		fun toJson() = buildJsonArray {
			entries.forEach { e ->
				add(
					buildJsonObject {
						put("id", e.id)
						put("textEn", e.text)
						put("textBn", e.text)
					},
				)
			}
		}
	}

	/**
	 * Correct value + up to three same-band distractors (same part of speech preferred, duplicate
	 * texts skipped), shuffled. The correct flag never enters the payload — only the answer key
	 * knows [Options.correctId].
	 */
	private fun mcqOptions(word: LocalVocabulary, pool: List<LocalVocabulary>, text: (LocalVocabulary) -> String): Options {
		val distractors = mutableListOf<LocalVocabulary>()
		val taken = mutableSetOf(text(word))
		// Two passes: same part of speech first, then anything still needed. A text only enters
		// `taken` when actually used, so pass 2 can still pick pass-1 rejects.
		for (samePosOnly in booleanArrayOf(true, false)) {
			for (candidate in pool) {
				if (distractors.size == MCQ_DISTRACTORS) break
				if (candidate.id == word.id ||
					(samePosOnly && candidate.partOfSpeech != word.partOfSpeech) ||
					!taken.add(text(candidate))
				) {
					continue
				}
				distractors.add(candidate)
			}
		}

		val correctId = UUID.randomUUID().toString()
		val entries = mutableListOf(OptionEntry(correctId, text(word)))
		distractors.forEach { entries.add(OptionEntry(UUID.randomUUID().toString(), text(it))) }
		entries.shuffle(random)
		return Options(entries, correctId)
	}

	private fun mcqKey(options: Options, revealText: String): JsonObject = buildJsonObject {
		put("correctOptionId", options.correctId)
		put("revealText", revealText)
	}

	companion object {
		/** SENTENCE_BUILD only for genuinely small sentences (word-bank stays tappable). */
		const val MAX_BUILD_TOKENS = 8
		private const val MIN_BUILD_TOKENS = 3
		private const val MCQ_DISTRACTORS = 3
	}
}
